GPS: add support for HIL_GPS

rmackay9 opened the pull request:
This PR adds support for the HIL_GPS message within the GPS MAV driver.  ROS's mavros apparently sends this message instead of GPS_INPUT so adding support improves integration with ROS.

This should resolve https://github.com/ArduPilot/ardupilot/issues/5344

rmackay9 commented:
By the way, I've asked "291" (https://github.com/291) in the Vision Projects room if he can test this.  https://gitter.im/ArduPilot/ardupilot/VisionProjects

WickedShell commented on libraries/AP_GPS/AP_GPS_MAV.cpp:
```diff
@@ -42,67 +42,107 @@ bool AP_GPS_MAV::read(void)
+            }
+            Vector3f vel(packet.vn/100.0f, packet.ve/100.0f, packet.vd/100.0f);
+            state.velocity = vel;
+            if (packet.cog < 65535) {
```
Technically should only be if ( packet.cog  < 36000) . Anything greater is out of spec for what we are expecting to have reported to us.

WickedShell commented on libraries/AP_GPS/AP_GPS_MAV.cpp:
```diff
@@ -42,67 +42,107 @@ bool AP_GPS_MAV::read(void)
+            loc.lat = packet.lat;
+            loc.lng = packet.lon;
+            if (have_alt) {
+                loc.alt = packet.alt;
```
I know this is not new but the packet.alt is in meters, and you are storing it to a Location struct, which isn't correct. (Either that or I'm way to tired). I think this one needs to be converted into cm's before storing. (Either that or the other one needs to come out of cm's).

WickedShell commented on libraries/AP_GPS/AP_GPS_MAV.cpp:
```diff
@@ -42,67 +42,107 @@ bool AP_GPS_MAV::read(void)
+            mavlink_msg_hil_gps_decode(msg, &packet);
+
+            state.time_week = 0;
+            state.time_week_ms  = packet.time_usec/1000;
```
If this doesn't increment in fixed/even steps I think that could be a problem for the EKF, (IE if it was 200ms then 180ms then 221 ms etc). Not sure what the sender code looks like, presumably its just locking GPS readings and passing them as they occurred so it would always count evenly.

WickedShell commented on libraries/AP_GPS/AP_GPS_MAV.cpp:
```diff
@@ -42,67 +42,107 @@ bool AP_GPS_MAV::read(void)
+            state.location.options = 0;
+
+            if (have_hdop) {
+                state.hdop = packet.hdop * 100; // convert to centimeters
```
Pedantic, but this is unit less, it's not in centimeters, not sure how pervasive the problem is, but since you had to tab the line in... :) (same with all other dop's)

WickedShell commented on libraries/AP_GPS/AP_GPS_MAV.cpp:
```diff
@@ -42,67 +42,107 @@ bool AP_GPS_MAV::read(void)
+            state.have_speed_accuracy = 0;
+            state.have_horizontal_accuracy = 0;
+            state.have_vertical_accuracy = 0;
+            state.num_sats = packet.satellites_visible;
```
Spec says that 255 means it's unknown, we probably don't want to be setting this to 255 as it would really mess with a user/auto gps swap algorthims.

WickedShell commented:
GPS_INPUT is a very new message, but it is a much better message then HIL_GPS. I know it won't happen instantly, but pushing the ROS side towards the better message is definitely worthwhile, we are just missing a lot of information in this message for assessing the accuracy of the GPS we are provided, which makes it harder for the arming checks/EKF to protect you from a bad fix, while the GPS_INPUT message can carry the needed accuracy information.

Other then the comments I made above (sorry for not making it 1 review and inadvertently spamming everyone with e-mail, made way more comments then I was anticipating...) seems reasonable and straightforward, but haven't tried to actually use it at all.

rmackay9 commented:
@WickedShell, thanks for the review.  I've added two commits that resolve the issues you've raised.  Catching the existing issue with the GPS_INPUT message was great, thanks!

Also rebased on master.
